### PR TITLE
Support for custom sql connection factory

### PR DESCRIPTION
**Pull request by tmasternak**
Part of #100 
Reason we need this: https://github.com/Particular/NServiceBus.SqlServer/pull/86
Waiting for: https://github.com/Particular/NServiceBus/pull/3158

### CODE REVIEW

**Comment by SimonCropp** [comment]
if we are giving them a delegate we dont need to pass them the connection string from config. they can read that in the delegate if they want to

**Comment by tmasternak** [comment]
How would they get it is such case?

**Comment by SimonCropp** [comment]
they own the connection string, either in code or in xml. no reason they should give it to use for us to give it back to them

**Comment by tmasternak** [comment]
That is true until we have multi-database scenario. In that case connection string passed to lambda will change. We do not have multi-database yet but that should happen sooner than later. In addition current signature is compliant with old configuration api.

**Comment by SimonCropp** [comment]
show me two code samples of how someone would configure 

1. all in code
2. in code combined with XML

as for multi database support what is the discriminator to switch between dbs?

**Comment by SimonCropp** [comment]
also if you are already making design decisions based on multidatabase show me some config scenarios on how that will work

**Comment by tmasternak** [comment]
I don't know how the scenarios will look in detail. That being said mapping transport addresses to connection strings (for multidatabase setup) and creating connection from a connection string are two different responsibilities and I think this should be reflected in extension points.

**Comment by SimonCropp** [comment]
so since we dont know what multidatabase will look like lets not try and design for it now.

so i am back to: if they are in control of creating a connection then there is no value in us passing them the connection string they have passed us. it is pointless

**Comment by tmasternak** [comment]
I think that we don't know the details but we do know something which influences the design.

To the second point. If user uses xml configuration he need to read the connection string one more time in lambda. Passing is more convenient and in addition we keep the old api unchanged.

**Comment by SimonCropp** [comment]
u r missing the point. if they are taking control of the connection creation it is entirely likely that they want to just have it hard coded, or perhaps read it from some other config like yaml or json.

and yes if they want to read it from xml they can write one line of code to do so. but we dont want to leak app.confix based logic into a code based api. we avoid doing this everywhere else in the core

**Comment by tmasternak** [comment]
Did you have a look at https://github.com/Particular/NServiceBus.SqlServer/pull/86? To my knowledge that is the main reason we need this.

File: src/NServiceBus.SqlServer/SqlServerTransport.cs [hunk 1]
```diff
@@ -33,15 +35,27 @@ SqlServerAddressProvider CreateAddressParser(ReadOnlySettings settings)
             return parser;
         }
 
+        SqlConnectionFactory CreateConnectoinFactory(string connectionString, ReadOnlySettings settings)
```
Comment (andreasohlund) [inline src/NServiceBus.SqlServer/SqlServerTransport.cs]: Typo

**Comment by SimonCropp** [comment]
a simple delegate will still support  #86

**Comment by tmasternak** [comment]
I agree the delegate would be fine. However what I'm pointing out is that self managing of connection string is not the reason this method was introduced.

**Comment by SimonCropp** [comment]
> However what I'm pointing out is that self managing of connection string is not the reason this method was introduced.

ok but a simple delegate still allows us to support 86 while not taking a contradictory stance to how other apis work.  It also means less code overall

**Comment by andreasohlund** [comment]
I'm with @SimonCropp here. I don't see the value of passing that string in, as Simon mentioned they can just read that them selfves if needed?

I'd say call YAGNI on this for now and handle this when we have more details on what we want to do with the multi data base support. (separate pull)

**Comment by tmasternak** [comment]
I've removed connection string from factory.

**Pull request merged**